### ISSUE TITLE
Are we doing the right thing with "Colorspace"?

### ISSUE THREAD

**Issue by Joshua-Ashton** [Plagman/gamescope#732]
I have not seen any difference with the "Colorspace" prop set vs without it.

Every monitor encountered seems to just assume BT2020 anyway when the PQ EOTF is selected in the HDR metadata.

There is also both DRM_MODE_COLORIMETRY_BT2020_RGB and DRM_MODE_COLORIMETRY_BT2020_YCC that we have to pick between.

**Comment by Joshua-Ashton** [comment]
Yes, we need to set `DRM_MODE_COLORIMETRY_BT2020_RGB`  or the display stays with sRGB/Rec.709 primaries.

Setting `DRM_MODE_COLORIMETRY_BT2020_YCC` doesn't make AMDGPU prefer to use YCC right now nor does that seem to be preferred by default or anything. It just makes the screen bright green at the moment.  :frog:
Seems like I can do at least 4K@120Hz + HDR right now (DSC obviously). 

Maybe there are some cases where link bandwidth going through certain passive/active docks would be lower and we might need to have chroma stuff going on...

Another thought, typically the driver is meant to perform any chroma subsamping, so it seems weird for _RGB or _YCC to be exposed to userspace at all.
Makes *somewhat* sense if you are outputting a chroma + luma planar format directly, but I didn't see anything hooked up for any form of subsampling demotion of the Colorspacefor RGB  -> YCC if that is meant to be happening right now in AMDGPU.

Rebased kernel colorspace stuff for testing:
https://gitlab.freedesktop.org/JoshuaAshton/linux-hdr
https://gitlab.freedesktop.org/JoshuaAshton/linux-hdr-pkgbuild

**Comment by swick** [comment]
> Setting DRM_MODE_COLORIMETRY_BT2020_YCC doesn't make AMDGPU prefer to use YCC right now nor does that seem to be preferred by default or anything

It shouldn't. It's just an infoframe for the display.

> Maybe there are some cases where link bandwidth going through certain passive/active docks would be lower and we might need to have chroma stuff going on... (or can just demote BPC?)

User space can set max bpc and drivers then lower bpc if the bandwidth requirements are too high for the current mode. They also might choose YCC formats to do the same. I'm a bit surprised that you don't seem to get YCC because the last information about amdgpu I have was that it defaults to it if possible. I really should look up the current state here or ask Harry.

Either way, actual bpc and ycc/rgb selection is currently completely driver defined.

> Currently AMDGPU has no notion of the EOTF of the display, so any subsampling or dithering (from bpc demotion) would probably look really bad as it would be done in the encoded PQ space rather than in linear space.

Dithering on linear values is counterproductive. You want the strength to be constant brightness, not constant luminance. A perceptual color space is optimal but dithering on electrical values is a good enough approximation.  Sub-sampling is also fine because you're doing so on the chroma channels alone.

> Really need to get a concept of per-plane and output TF hooked up.

We have been pushing back hard against putting high level details such as TFs into the kernel for various reasons. https://gitlab.freedesktop.org/pq/color-and-hdr/-/issues/11 and the links in there should be able to shed some light on why. The tl;dr is non-standard/profiled monitors and avoiding differences between KMS pipeline and shader pipeline for on-demand KMS offloading.

**Comment by Joshua-Ashton** [comment]
> Dithering on linear values is counterproductive. You want the strength to be constant brightness, not constant luminance. A perceptual color space is optimal but dithering on electrical values is a good enough approximation. Sub-sampling is also fine because you're doing so on the chroma channels alone.

Makes sense for chroma, but I don't understand the dithering part.

I am applying the same line of thinking I would to lerping/filtering like a game would here.
All games work in linear space and then convert to PQ (or just leave it if it's scRGB).
Seeing how horrible blending with an sRGB/PQ encoded value ends up being, I don't see how dithering with PQ curve encoded values is meant to be any better, when the end result is meant to result is meant to effectively emulate a blend between two values?

**Comment by Joshua-Ashton** [comment]
> We have been pushing back hard against putting high level details such as TFs into the kernel for various reasons. 

I would say we definitely have a difference of opinion here but for different reasons.
On Deck and other portable devices, it's important for us to take advantage of the hardware as we are very limited by both power and bandwidth.
Doing such would involve using the in-built transfer function LUTs in the hardware. Not taking advantage of these blocks is actually quite detrimental in terms of power/perf in our area.
I would also argue that having an enumeration of TFs in the DRM interface and defining numerically are not mutually exclusive concepts. We can do both and reap the benefits of both sides.
I don't think we would be open to leaving the HW blocks with the in-built LUTs on the table.

**Comment by swick** [comment]
Anyway, good to know that some displays at least behave as expected. Choosing the RGB variant of the Colorspace might work for you but in general there is no guarantee it's the right one and we need to either let user space control RGB/YCC or let the kernel select the right variant. Having control over subsampling in user space would be nicer because YCC is undesirable in some situations.

> I am applying the same line of thinking I would to lerping/filtering like a game would here.

Dithering makes most sense if you ignore the color aspect and think about it as a signal processing issue of error diffusion. There is really no need to involve color at all, except that HVS integrates over an area.


> I don't think we would be open to leaving the HW blocks with the in-built LUTs on the table.

We absolute want to use the color pipeline of scanout engines. What we don't want is a high level description of "this is the content, this is the output, do some magic to convert the one to the other". Instead we want to expose all the operations in the color pipeline. If there is a 1D LUT somewhere in the pipeline, we expose the 1D LUT, if there is a block which encodes EOTFs, we expose it. Which pipeline should be used and how it should be configured is then up to user space.

**Comment by Joshua-Ashton** [comment]
Had a user report this issue with out current setup:

![green](https://user-images.githubusercontent.com/21316711/211283994-af2ad7e8-6f63-43b9-a56c-7f8ab095e25b.png)

I had them change BT2020_RGB to BT2020_YCC which didn't fix the issue which was interesting.

**Comment by Joshua-Ashton** [comment]
> I'm a bit surprised that you don't seem to get YCC because the last information about amdgpu I have was that it defaults to it if possible. I really should look up the current state here or ask Harry.

I believe it only does this on HDMI, which may be why I am seeing such weirdness on HDMI.
I am going to experiment mode.

```c
	if (stream->signal == SIGNAL_TYPE_HDMI_TYPE_A) {
		if (!adjust_colour_depth_from_display_info(timing_out, info) &&
		    drm_mode_is_420_also(info, mode_in) &&
		    timing_out->pixel_encoding != PIXEL_ENCODING_YCBCR420) {
			timing_out->pixel_encoding = PIXEL_ENCODING_YCBCR420;
			adjust_colour_depth_from_display_info(timing_out, info);
		}
	}
```

**Comment by Joshua-Ashton** [comment]
This was fixed with my following commits:

![image](https://user-images.githubusercontent.com/21316711/211841842-e495d7db-d891-486e-b37b-f59cf3a0047a.png)

The "Colorspace" property was missing the implicit demotion from BT2020_RGB -> BT2020_YCBCR that happens for typical sRGB.

On top of this, the YCBCR2020 matrix in AMDGPU had a bad fourth column, where it was just zero'ed out, which was making things more green as the bias from [-1,1] -> [0,1] for chroma was missing.

**Comment by misyltoad** [comment]
I remember discussing this before, and we agreed that the colorimetry should not care about the encoding and the rgb/ycc/ycbcr options should all automatically be fixed up to the right one for auto pixel encoding.

**Comment by darksylinc** [comment]
The problem about that is that it assumes the monitor EDID isn't broken.

I suspect a better model is an enum where one option in the color space just leaves it to the driver to autoselect, while another option forces the encoding.
